packages/runtime/container-runtime/src/garbageCollection.ts

```diff
@@ -0,0 +1,244 @@
+const GCVersion = 1;
+
+// Local storage key to turn GC on / off.
+const runGCKey = "FluidRunGC";
+// Local storage key to turn GC test mode on / off.
+const gcTestModeKey = "FluidGCTestMode";
+// Local storage key to turn GC sweep on / off.
+const runSweepKey = "FluidRunSweep";
+
+/** The used state statistics of a node. */
+export interface IUsedStateStats {
+    totalNodeCount: number;
+    unusedNodeCount: number;
+}
+
+/** The statistics of the system state after a garbage collection run. */
+export interface IGCStats {
+    totalNodes: number;
+    deletedNodes: number;
+    totalDataStores: number;
+    deletedDataStores: number;
+}
+
+/** Defines the APIs for the runtime object to be passed to the garbage collector. */
+export interface IGarbageCollectionRuntime {
+    /** Returns the garbage collection data of the runtime. */
+    getGCData(fullGC?: boolean): Promise<IGarbageCollectionData>;
+    /** After GC has run, called to notify the runtime of routes that are used in it. */
+    updateUsedRoutes(usedRoutes: string[]): IUsedStateStats;
+    /** After GC has run, called to delete objects in the runtime whose routes are unused. */
+    deleteUnusedRoutes(unusedRoutes: string[]): void;
+}
+
+/** Defines the contract for the garbage collector. */
+export interface IGarbageCollector {
+    /** The GC version that was used to generate the GC data to be written in a summary. */
+    gcSummaryFeatureVersion: number;
+    /** Tells whether GC should run or not. */
+    shouldRunGC: boolean;
+    /** Tells whether the GC version has changed compared to the version in the latest summary. */
+    hasGCVersionChanged: boolean;
+    /** Run garbage collection and update the reference / used state of the system. */
+    collectGarbage(
+        options: { logger?: ITelemetryLogger, runGC?: boolean, runSweep?: boolean, fullGC?: boolean },
+    ): Promise<IGCStats>;
+    /** Called when the latest summary of the system has been refreshed. */
+    latestSummaryStateRefreshed(result: RefreshSummaryResult, readAndParseBlob: ReadAndParseBlob): Promise<void>;
+}
+
+/**
+ * The garbage collector for the container runtime. It consolidates the garbage collection functionality and maintains
+ * its state across summaries.
+ */
+export class GarbageCollector implements IGarbageCollector {
+    public readonly gcEnabled: boolean;
+    public readonly shouldRunGC: boolean;
+    private readonly shouldRunSweep: boolean;
+    private readonly testMode: boolean;
+    private readonly logger: ITelemetryLogger;
+
+    // The current GC version that this container is running.
+    private readonly currentGCVersion = GCVersion;
+    // This is the version of GC data in the latest summary being tracked.
+    private latestSummaryGCVersion: GCVersion;
+
+    /**
+     * Returns the GC version that was used to generate the GC data to be written in a summary.
+     */
+    public get gcSummaryFeatureVersion(): number {
```

Comment:
Does it compile, giving that property is not readonly on interface, and there is no setter?

Reply:
Yeah, it does compile!